File: diumoo/core/DMNotificationCenter/DMNotificationCenter.m
//
//  DMNotificationCenter.m
//  diumoo
//
//  Created by Shanzi on 12-7-22.
//  Copyright (c) 2012年 __MyCompanyName__. All rights reserved.
//

#import "DMNotificationCenter.h"

@implementation DMNotificationCenter

#pragma init and dealloc

-(id) init
{
    if (self = [super init]) {
        if (NSClassFromString(@"NSUserNotification")) {
            [[NSUserNotificationCenter defaultUserNotificationCenter] setDelegate:self];
        }
    }
    return self;
}

-(void) dealloc
{
    [[NSUserNotificationCenter defaultUserNotificationCenter] setDelegate:nil];
}

#pragma ---

-(void) notifyMusicWithItem:(DMPlayableItem *)item
{
    NSUserDefaults* values = [NSUserDefaults standardUserDefaults];
    
    if ([[values valueForKey:@"enableGrowl"] integerValue] == NSOnState)
    {
        NSString* detail = [NSString stringWithFormat:@"%@ - <%@>",item.musicInfo[@"artist"],item.musicInfo[@"albumtitle"]];
        
        NSUserNotification *notification = [[NSUserNotification alloc] init];
        NSUserNotificationCenter *center = [NSUserNotificationCenter defaultUserNotificationCenter];
        notification.title = item.musicInfo[@"title"];
        notification.contentImage = item.cover;
        notification.informativeText = detail;
        notification.soundName = nil;
        [center deliverNotification: notification];
    }
    
    if([[values valueForKey:@"displayAlbumCoverOnDock"] integerValue]==NSOnState){
        [NSApp setApplicationIconImage:item.cover];
    }
}

-(void) notifyBitrate
{
    NSLog(@"Bitrate changed Notification");
    NSUserDefaults* values = [NSUserDefaults standardUserDefaults];
    
    if ([[values valueForKey:@"enableGrowl"] integerValue] == NSOnState)
    {
        NSString* title = NSLocalizedString(@"BITRATE_CHANGED", nil);
        NSString* detail = [NSLocalizedString(@"BITRATE_CHANGED_TO_VALUE", nil)
                            stringByAppendingFormat:@"%@ Kbps",[values valueForKey:@"musicQuality"]];
        
        NSUserNotification *notification = [[NSUserNotification alloc] init];
        NSUserNotificationCenter *center = [NSUserNotificationCenter defaultUserNotificationCenter];
        notification.title = title;
        notification.informativeText = detail;
        notification.soundName = nil;
        [center deliverNotification: notification];
    }
    
}

- (void) copylinkNotification:(NSString *)url
{
    NSUserDefaults* values = [NSUserDefaults standardUserDefaults];
    if ([[values valueForKey:@"enableGrowl"] integerValue] == NSOnState)
    {
        NSString* title = NSLocalizedString(@"SHARE_LINK_TITLE", nil);
        NSString* detail = url;
        
        NSUserNotification *notification = [[NSUserNotification alloc] init];
        NSUserNotificationCenter *center = [NSUserNotificationCenter defaultUserNotificationCenter];
        notification.title = title;
        notification.informativeText = detail;
        notification.soundName = nil;
        [center deliverNotification: notification];
    }
}

-(void) clearNotifications
{
    if(NSClassFromString(@"NSUserNotification")){
        [[NSUserNotificationCenter defaultUserNotificationCenter] removeAllDeliveredNotifications];
    }
}

//this method forced Notification Center present notification whatever the application is foreground
- (BOOL)userNotificationCenter:(NSUserNotificationCenter *)center shouldPresentNotification:(NSUserNotification *)notification
{
    return YES;
}



@end

File: diumoo/core/DMNotificationCenter/DMNotificationCenter.swift
//
//  DMNotificationCenter.swift
//  diumoo
//
//  Created by Yancheng Zheng on 6/18/16.
//
//

import Foundation
import AppKit

public class DMNotificationCenter : NSObject, NSUserNotificationCenterDelegate  {
 
    internal let pref = UserDefaults.standard()
    internal let NCCenter = NSUserNotificationCenter.default()
    internal let needToUpdateDock : Bool
    
    public func userNotificationCenter(_ center: NSUserNotificationCenter, shouldPresent notification: NSUserNotification) -> Bool {
        return true
    }
    
    override init() {
        // Show Dock icon setting won't change without restart app, so it's safe to put it here
        // I'm not sure about how expensive this is
        needToUpdateDock = (Int(self.pref.value(forKey: "displayAlbumCoverOnDock") as! NSNumber) == NSOnState)
        super.init()
        NSUserNotificationCenter.default().delegate = self
    }
    
    deinit {
        NSUserNotificationCenter.default().delegate = nil
    }
    
    func isNotificationON() -> Bool {
        return Int(self.pref.value(forKey: "enableNotification") as! NSNumber) == NSOnState
    }
    
    public func notifyMusicPlayback(withItem item: DMPlayableItem) {
        if isNotificationON() {
            let detail = String("\(item.musicInfo["artist"]!) - <\(item.musicInfo["albumtitle"]!)>")
            let aNotification = NSUserNotification.init()
            aNotification.title = item.musicInfo["title"] as? String
            aNotification.informativeText = detail
            aNotification.contentImage = item.cover
            aNotification.soundName = nil
            self.NCCenter.deliver(aNotification)
        }
        
        if self.needToUpdateDock {
            NSApplication.shared().applicationIconImage = item.cover
        }
    }
    
    public func notifyBitrate() {
        if isNotificationON() {
            let title = NSLocalizedString("BITRATE_CHANGED", comment: "Music bitrate changed.")
            var detail = NSLocalizedString("BITRATE_CHANGED_TO_VALUE", comment: "Music bitrate changed to.")
            let quality = pref.value(forKey: "musicQuality") as? Int
            detail.append(String("\(quality!) Kbps"))
            
            let aNotification = NSUserNotification.init()
            aNotification.title = title
            aNotification.informativeText = detail
            aNotification.soundName = nil
            
            self.NCCenter.deliver(aNotification)
            
        }
    }
    
    public func clearNotifications() {
        self.NCCenter.removeAllDeliveredNotifications()
    }
    
    public func copylinkNotification (URLStr: String) {
        if isNotificationON() {
            let title = NSLocalizedString("SHARE_LINK_TITLE", comment: "share link")
            
            let aNotification = NSUserNotification.init()
            
            aNotification.title = title
            aNotification.informativeText = URLStr
            aNotification.soundName = nil
            
            self.NCCenter.deliver(aNotification)
        }
        
    }
}
